test/extensions/filters/http/lua/wrappers_test.cc

```diff
@@ -418,6 +418,41 @@ TEST_F(LuaStreamInfoWrapperTest, SetGetAndIterateDynamicMetadata) {
   wrapper.reset();
 }
 
+// Verify that binary values could also be extracted from dynamicMetadata().
+TEST_F(LuaStreamInfoWrapperTest, GetDynamicMetadataBinaryData) {
+  const std::string SCRIPT{R"EOF(
+    function callMe(object)
+      local metadata = object:dynamicMetadata():get("envoy.pp")
+      local bin_data = metadata["bin_data"]
+      local str_table = { }
+      local data_length = string.len(metadata["bin_data"])
+      for idx = 1, data_length do
+        if string.byte(bin_data, idx) ~= 0 then
+          str_table[#str_table + 1] = string.char(string.byte(bin_data, idx))
+        end
+      end
+      testPrint('Data: ' .. table.concat(str_table, ''))
+    end
+  )EOF"};
+
+  ProtobufWkt::Value metadata_value;
+  constexpr uint8_t buffer[] = {'h', 'e', 0x00, 'l', 'l', 'o'};
+  metadata_value.set_string_value(reinterpret_cast<char const*>(&buffer), sizeof(buffer));
+  ProtobufWkt::Struct metadata;
+  metadata.mutable_fields()->insert({"bin_data", metadata_value});
+
+  setup(SCRIPT);
+
+  StreamInfo::StreamInfoImpl stream_info(Http::Protocol::Http2, test_time_.timeSystem(), nullptr);
+  (*stream_info.metadata_.mutable_filter_metadata())["envoy.pp"] = metadata;
+  Filters::Common::Lua::LuaDeathRef<StreamInfoWrapper> wrapper(
+      StreamInfoWrapper::create(coroutine_->luaState(), stream_info), true);
+
+  EXPECT_CALL(printer_, testPrint("Data: hello"));
```

Comment:
Ditto in the test above, but we want to see the 0x00 reflected in our assertion somehow. I recommend the hex escape codes you were using before as they make it pretty clear that everything stayed the same byte by byte, but other methods could work too. This assertion though makes me wonder why we dropped the 0x00, is it your lua code or is it a bug that dropped the 0x00 even before your lua code.

Reply:
Changed it to hex instead. Thanks!